lib/src/solver/incompatibility.dart

```diff
@@ -44,39 +52,84 @@ class Incompatibility {
       }
     }
 
-    return new Incompatibility._(byName.values.expand((byRef) {
-      // If there are any positive terms for a given package, we can discard
-      // any negative terms.
-      var positiveTerms =
-          byRef.values.where((term) => term.isPositive).toList();
-      if (positiveTerms.isNotEmpty) return positiveTerms;
+    return new Incompatibility._(
+        byName.values.expand((byRef) {
+          // If there are any positive terms for a given package, we can discard
+          // any negative terms.
+          var positiveTerms =
+              byRef.values.where((term) => term.isPositive).toList();
+          if (positiveTerms.isNotEmpty) return positiveTerms;
 
-      return byRef.values;
-    }).toList());
+          return byRef.values;
+        }).toList(),
+        cause);
   }
 
-  Incompatibility._(this.terms);
+  Incompatibility._(this.terms, this.cause);
 
   String toString() {
+    if (cause == IncompatibilityCause.dependency) {
+      assert(terms.length == 2);
+
+      var depender = terms.first;
+      var dependee = terms.last;
+      assert(depender.isPositive);
+      assert(!dependee.isPositive);
+
+      if (depender.constraint.isAny) {
+        return "all versions of ${_terseRef(depender)} "
+            "depend on ${dependee.package.toTerseString()}";
+      } else {
+        return "${depender.package.toTerseString()} depends on "
+            "${dependee.package.toTerseString()}";
+      }
+    } else if (cause == IncompatibilityCause.sdk) {
+      assert(terms.length == 1);
+      assert(terms.first.isPositive);
+
+      if (terms.first.constraint.isAny) {
+        return "no versions of ${_terseRef(terms.first)} "
+            "are compatible with the current SDK";
+      } else {
+        return "${terms.first.package.toTerseString()} is incompatible with "
+            "the current SDK";
+      }
```

Comment:
Might be useful to show the SDK version number here, no?

Reply:
Added a TODO.